tests/test_packages/test_skills/test_generic_buyer/test_behaviours.py

```diff
@@ -0,0 +1,389 @@
+        fipa_dialogue = FipaDialogue(
+            DialogueLabel(("1", "1"), "counterparty", "me"),
+            "me",
+            FipaDialogue.Role.BUYER,
+        )
+        fipa_dialogues = self.skill.skill_context.fipa_dialogues
+
+        # before
+        for end_state_numbers in fipa_dialogues.dialogue_stats.self_initiated.values():
+            assert end_state_numbers == 0
+        for end_state_numbers in fipa_dialogues.dialogue_stats.other_initiated.values():
+            assert end_state_numbers == 0
+
+        # operation
+        fipa_handler._handle_decline(incoming_message, fipa_dialogue, fipa_dialogues)
+
+        # after
+        for end_state_numbers in fipa_dialogues.dialogue_stats.other_initiated.values():
+            assert end_state_numbers == 0
+        for (
+            end_state,
+            end_state_numbers,
+        ) in fipa_dialogues.dialogue_stats.self_initiated.items():
+            if end_state == FipaDialogue.EndState.DECLINED_CFP:
+                assert end_state_numbers == 1
+            else:
+                assert end_state_numbers == 0
+
+    def test_fipa_handler_handle_decline_decline_accept(self):
+        """Test the _handle_decline method of the fipa handler where the end state is decline_accept."""
+        # setup
+        fipa_handler = cast(GenericFipaHandler, self.skill.skill_context.handlers.fipa)
+
+        incoming_message = FipaMessage(
+            dialogue_reference=("1", "1"),
+            message_id=4,
+            target=3,
+            performative=FipaMessage.Performative.DECLINE,
+        )
+        incoming_message.sender = "counterparty"
+        incoming_message.to = "me"
+        fipa_dialogue = FipaDialogue(
+            DialogueLabel(("1", "1"), "counterparty", "me"),
+            "me",
+            FipaDialogue.Role.BUYER,
+        )
+        fipa_dialogues = self.skill.skill_context.fipa_dialogues
+
+        # before
+        for end_state_numbers in fipa_dialogues.dialogue_stats.self_initiated.values():
+            assert end_state_numbers == 0
+        for end_state_numbers in fipa_dialogues.dialogue_stats.other_initiated.values():
+            assert end_state_numbers == 0
+
+        # operation
+        fipa_handler._handle_decline(incoming_message, fipa_dialogue, fipa_dialogues)
+
+        # after
+        for end_state_numbers in fipa_dialogues.dialogue_stats.other_initiated.values():
+            assert end_state_numbers == 0
+        for (
+            end_state,
+            end_state_numbers,
+        ) in fipa_dialogues.dialogue_stats.self_initiated.items():
+            if end_state == FipaDialogue.EndState.DECLINED_ACCEPT:
+                assert end_state_numbers == 1
+            else:
+                assert end_state_numbers == 0
+
+    def test_fipa_handler_handle_match_accept(self):
+        """Test the _handle_match_accept method of the fipa handler."""
+        # ToDo does not work; fails on dialogues.update(). Investigate why.
```

Comment:
This test does not work (failing on dialogues.update() line 363).

Reply:
It also shows that for some unit tests, a lot of setup is needed to get some part of a skill code working. For instance to get `handle_match_accept` working, you need to pass it a dialogue containing all other messages prior to the `match_accept`  (cfp, propose, accept).

Reply:
This highlights the issue that for complex skills, to unit test a part of the code that depends on a large state, that whole state has to be replicated to get the code working. 

Maybe we can create something to make that easy? For this test, an easy way of populating a dialogue object, and given heavy reliance on dialogues, this would be useful across tests.